### PR TITLE
docs(Cast): update to reflect new design

### PR DESCRIPTION
**Pull request by varunsrin**
Update Section 4.1 to cover the latest designs for casts

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -203,107 +203,78 @@ All messages must pass the following validations in addition to specific validat
...
 
-### 4.1.1 Short Text Casts
+A Cast can be added with a `CastAdd` message which is placed in the CRDT's **add-set**. Each cast is indexed by its hash which is guaranteed to be unique unless the casts are identical. By extension, two add messages can never conflict unless they are identical, in which case one can be discarded safely. 
 
-A _Short Text Cast_ is a 280 character public message created by an account. It can stand on its own or as a reply if `parentUri` points to another cast, on-chain item or URL. When deleted, a Short Text Cast should be soft-deleted and its content removed, but its replies are still considered valid and can be displayed to the user.
+A Cast can be removed with a `CastRemove` message which contains a reference to the target `CastAdd`'s hash. When received, the target is removed from the add-set if present and the remove is added to the rem-set. Conflicts between adds and removes are handled with Remove-Wins runes and conflicts between removes are handled with Last-Write-Wins rules, falling back to lexicographical ordering in case of a tie.
```
Comment (pfletcherhill) [inline README.md:210]: "Remove-Wins rules" vs "runes"

File: README.md [hunk 2]
```diff
@@ -203,107 +203,78 @@ All messages must pass the following validations in addition to specific validat
 
 ## 4.1 Casts
 
-A _Cast_ is a public message created by a user that is displayed on their profile. It can contain text and links to media, on-chain activity or other casts. Users are allowed to delete casts at any time. Casts come in several different flavors and the protocol can be extended to support many more types in the future. Each type has its own replication mechanisms which are defined below.
+A _Cast_ is a public message created by a user which contains text and can also embed media, on-chain activity or evens other casts. Casts are stored in a two-phase set CRDT[^two-phase-set] that handles conflict resolution.
 
-### 4.1.1 Short Text Casts
+A Cast can be added with a `CastAdd` message which is placed in the CRDT's **add-set**. Each cast is indexed by its hash which is guaranteed to be unique unless the casts are identical. By extension, two add messages can never conflict unless they are identical, in which case one can be discarded safely.
```
Comment (pfletcherhill) [inline README.md:208]: Annoying nit: "add-set" is bolded here but "rem-set" is not bolded in the next paragraph

File: README.md [hunk 3]
```diff
@@ -203,107 +203,78 @@ All messages must pass the following validations in addition to specific validat
...
-1. If there is an `a` in the add-set where `a.hash` equals `d.targetCastHash`, delete it.
-2. If there is an `r` in the rem-set where `r.targetCastHash` equals `d.targetCastHash`
-   - If `r > d`, discard `d`
-   - If `r < d`, delete `r` and add `d` into the rem-set
-3. Otherwise, add `d` to the rem-set.
-
-### 4.1.2 Recasts
-
-A _Recast_ expresses an intent to share another cast. An account can recast any cast, including its own, but may not recast the same cast multiple times.
+A _Cast Add_ can contain up to 320 characters of unicode text and two URI's that can have up to 256 characters. Clients are responsible for unpacking and rendering the URI's along with the text.
```
Comment (pfletcherhill) [inline README.md:214]: Should this be "URLs" instead of "URI's" ?

Comment (pfletcherhill) [reply README.md:214]: And what about mentions?

File: README.md [hunk 4]
```diff
@@ -203,107 +203,78 @@ All messages must pass the following validations in addition to specific validat
...
-### 4.1.2 Recasts
-
-A _Recast_ expresses an intent to share another cast. An account can recast any cast, including its own, but may not recast the same cast multiple times.
+A _Cast Add_ can contain up to 320 characters of unicode text and two URI's that can have up to 256 characters. Clients are responsible for unpacking and rendering the URI's along with the text. 
 
 ```ts
-type CastRecast = {
-  targetCastUri: URI;
-  schema: 'farcaster.xyz/schemas/v1/cast-recast';
+type CastAddMessage = {
```
Comment (pfletcherhill) [inline README.md:217]: `CastAddBody`

File: README.md [hunk 5]
```diff
@@ -203,107 +203,78 @@ All messages must pass the following validations in addition to specific validat
...
 
-Recasts can be stored using a two-phase set just like Short Text Casts. The remove operation is identical, while the addition operation is performed with `CastRecast`, and has one additional rule:
+A Cast message must pass the following validation steps:
 
-1. If there is a CastRecast `a` in the add **add-set** with the same `targetCastUri` and `fid` as the incoming CastRecast `rc`:
-   - If `a > rc`, discard `d`
-   - If `a < rc`, delete `a` and add `rc` into the rem-set
+1. `text` must contain <= 320 valid unicode characters
+2. `embed` must contain between 0 and 2 `items`
+3. `item` must be a URI of at most 256 characters
```
Comment (pfletcherhill) [inline README.md:245]: What's item?

File: README.md [hunk 6]
```diff
@@ -203,107 +203,78 @@ All messages must pass the following validations in addition to specific validat
...
+4. `parent`, if present, must be a valid URI not equal to this message's URI
```
Comment (pfletcherhill) [inline README.md:246]: Worth adding a mention of what one for a cast would look like? (i.e. `fid:<fid>/cast:<hash>`)

File: README.md [hunk 7]
```diff
@@ -203,107 +203,78 @@ All messages must pass the following validations in addition to specific validat
...
 
-### 4.1.3 Remove Messages
+### 4.1.2 Remove Messages
 
-A remove instructs the set to remove a previously created cast. It is a type of soft delete where the content of the message is removed but its hash is retained forever. A user who has a copy of the remove message can prove that the message was posted at some point by the author.
+A _Cast Remove_ only contains a reference to the hash of the _Cast Add_. It allows for permanent deletion of casts while eliding the data of the original cast.
```
Comment (pfletcherhill) [inline README.md:250]: TIL that "elide" is a word

File: README.md [hunk 8]
```diff
@@ -203,107 +203,78 @@ All messages must pass the following validations in addition to specific validat
...
+type CastRemoveBody = {
+  hash: string;
 };
 ```
 
-The remove message must contain the hash of the cast being removed and omit all the other properties. The set can then remove the cast and its contents from the network, which is desirable from a user perspective. The set ensures that the remove message does not reference itself, and that it is only applied as a remove operation if it has a timestamp higher than that of the message it references. The rules for validation are described below, but the rules for merging are specified in each set that implements the remove operation.
+The message must pass the following validation steps: 
 
-<!-- Diagram showing the remove operations -->
+1. `message.data.body.hash` must not equal the `message.envelope.hash`.
```
Comment (pfletcherhill) [inline README.md:260]: I think just including the rule for `hash` is fine. The others are covered by generic message validation rules.

**Review by pfletcherhill — APPROVED** [review]
Approving, but left some comments with typos and nits for you to address before merging

File: README.md (hunk 5 above)
Comment (varunsrin) [reply README.md:245]: see `embed` object above

File: README.md (hunk 7 above)
Comment (varunsrin) [reply README.md:250]: i learned it from rustlang

File: README.md (hunk 8 above)
Comment (varunsrin) [reply README.md:260]: I'm not sure I understand the suggestion

Comment (pfletcherhill) [reply README.md:265]: The docs include the validation for timestamp and fid in another section. So the only validation that's specific to CastRemove messages is the rule about the hash not referencing itself.

Comment (varunsrin) [reply README.md:265]: ah yeah, i might rework that so leaving this in here for now as i juggle PRs

**Pull request merged**